src/JSInterop/Microsoft.JSInterop/src/DotNetObjectRefManager.cs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.JSInterop
+{
+    internal class DotNetObjectRefManager
+    {
+        private readonly object _storageLock = new object();
+        private long _nextId = 1; // Start at 1, because 0 signals "no object"
+        private readonly Dictionary<long, IDotNetObjectRef> _trackedRefsById = new Dictionary<long, IDotNetObjectRef>();
+        private readonly Dictionary<IDotNetObjectRef, long> _trackedIdsByRef = new Dictionary<IDotNetObjectRef, long>();
+        public static DotNetObjectRefManager Current
+        {
+            get
+            {
+                if (!(JSRuntime.Current is JSRuntimeBase jsRuntimeBase))
+                {
+                    throw new InvalidOperationException("JSRuntime must be set up to use DotNetObjectRef.");
+                }
+
+                return jsRuntimeBase.ObjectRefManager;
+            }
+        }
+
+        public long TrackObject(IDotNetObjectRef dotNetObjectRef)
+        {
+            lock (_storageLock)
+            {
+                // Assign an ID only if it doesn't already have one
```

Comment:
I know you're inheriting this logic from the earlier version, but thinking about it now, this seems wrong. It's possible for two unrelated code paths to wrap the same object instance and pass it to JS as a `DotNetObject`. The two unrelated code paths will eventually dispose the reference.

But this will lead to a conflict, won't it? After the first receipient disposed the reference, the other may still be trying to use it.

So, it seems like we should always create a new wrapper and ID on each call.

Reply:
If this reasoning is correct, then we shouldn't need `_trackedIdsByRef` at all either.